server/src/main/java/io/crate/analyze/UpdateAnalyzer.java

```diff
@@ -159,70 +169,120 @@ private LinkedHashMap<Reference, Symbol> getAssignments(List<Assignment<Expressi
+                        normalizer.normalize(sourceExprAnalyzer.convert(assignment.expression(), exprCtx), txnCtx),
+                        arrayElementRefForValueNormalization,
+                        tableInfo,
+                        s -> normalizer.normalize(s, txnCtx));
 
-            Symbol source = ValueNormalizer.normalizeInputForReference(
-                normalizer.normalize(sourceExprAnalyzer.convert(assignment.expression(), exprCtx), txnCtx),
-                targetCol,
-                tableInfo,
-                s -> normalizer.normalize(s, txnCtx)
-            );
-            if (assignmentByTargetCol.put(targetCol, source) != null) {
-                throw new IllegalArgumentException("Target expression repeated: " + targetCol.column().sqlFqn());
+                    arraySetFunctionAllocator.put(targetCol, indexToUpdate, targetValue);
+                    continue;
+                } else if (unwrapReferenceFromCast(baseCol) instanceof Reference targetCol) {
+                    rejectUpdatesToFieldsOfObjectArrays(tableInfo, targetCol, IS_OBJECT_ARRAY);
+                }
+            }
+            throw new IllegalArgumentException(
+                String.format(
+                    Locale.ENGLISH,
+                    "cannot use expression %s as a left side of an assignment",
+                    assignment.columnName()));
+        }
+        for (var e : arraySetFunctionAllocator.allocate().entrySet()) {
+            if (assignmentByTargetCol.put(e.getKey(), e.getValue()) != null) {
+                throw new IllegalArgumentException("Target expression repeated: " + e.getKey().column().sqlFqn());
             }
         }
         return assignmentByTargetCol;
     }
 
-    private static boolean hasMatchingParent(TableInfo tableInfo, Reference info, Predicate<Reference> parentMatchPredicate) {
+    private static void rejectUpdatesToFieldsOfObjectArrays(TableInfo tableInfo, Reference info, Predicate<Reference> parentMatchPredicate) {
         for (var parent : tableInfo.getParents(info.column())) {
             if (parentMatchPredicate.test(parent)) {
-                return true;
+                throw new IllegalArgumentException("Updating fields of object arrays is not supported");
             }
         }
-        return false;
     }
 
-    private static class AssignmentNameValidator extends AstVisitor<Void, Boolean> {
+    private static class ArraySetFunctionAllocator {
+        private final Map<Reference, Map<Symbol, Symbol>> mappings;
+        private final BiFunction<String, List<Symbol>, Symbol> functionAllocator;
 
-        private static final AssignmentNameValidator INSTANCE = new AssignmentNameValidator();
-
-        static void ensureNoArrayElementUpdate(Expression expression) {
-            expression.accept(INSTANCE, false);
+        public ArraySetFunctionAllocator(BiFunction<String, List<Symbol>, Symbol> functionAllocator) {
```

Comment:
```suggestion
        public ArraySetFunctionAllocator(BiFunction<String, List<Symbol>, Symbol> allocateFunction) {
```

For `Function` / `Predicate` interfaces I find it helps to use a verb instead of a noun - same as you'd name a function. It gives more meaning to the `apply` - you apply the function.
`allocateFunction.apply(arg)` vs. `functionAllocator.apply`

But just a suggestion.